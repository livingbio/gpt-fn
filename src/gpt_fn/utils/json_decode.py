
from functools import wraps

def state(fn):
    @wraps(fn)
    def wrapper(input: str, stack: list[str] = []) -> str:
        r = fn(input, stack)
        if r is None:
            raise ValueError("Invalid JSON {input} in {fn.__name__}")
        return r
    return wrapper

def state_start(input: str) -> str:
    return state_root_object(input, ["$"])

@state
def state_root_object(input: str, stack: list[str]) -> str:
    input = input.strip()

    if input[0] == "{":
        return input[0] + state_object(input[1:], stack + ["{"])
    
@state
def state_finish(input: str, stack: list[str]) -> str:
    input = input.strip()

    if not input:
        return ""

@state
def state_object(input: str, stack: list[str]) -> str:
    input = input.strip()

    if input[0] == "}" and stack[-1] == "{":
        return input[0] + state_post_object(input[1:], stack[:-1])

    if input[0] == '"':
        return input[0] + state_property_string(input[1:], stack)

@state
def state_post_object(input: str, stack: list[str]) -> str:
    if stack[-1] in {"{", "["}:
        return state_post_value(input, stack)

    if stack[-1] == "$":
        return state_finish(input, stack)
    
@state
def state_post_property(input: str, stack: list[str]) -> str:
    input = input.strip()

    if input[0] == ":":
        return input[0] + state_value(input[1:], stack)

@state
def state_value(input: str, stack: list[str]) -> str:
    input = input.strip()

    if input[0] == "f":
        assert input[:5] == "false"
        return input[:5] + state_post_value(input[5:], stack)
    if input[0] == "t":
        assert input[:4] == "true"
        return input[:4] + state_post_value(input[4:], stack)
    if input[0] == "n":
        assert input[:4] == "null"
        return input[:4] + state_post_value(input[4:], stack)
    if input[0].isdigit() or input[0] == "-":
        return input[0] + state_int(input, stack)
    if input[0] == '"':
        return input[0] + state_value_string(input[1:], stack)
    if input[0] == "{":
        return input[0] + state_object(input[1:], stack + ["{"])
    if input[0] == "[":
        return input[0] + state_value(input[1:], stack + ["["]) 
    if input[0] == "]":
        assert stack[-1] == "["
        return input[0] + state_post_value(input[1:], stack[:-1])
    
@state
def state_post_value(input: str, stack: list[str]) -> str:
    input = input.strip()

    if input[0] == ",":
        if stack[-1] == "[":
            return input[0] + state_value(input[1:], stack)
        if stack[-1] == "{":
            return input[0] + state_object(input[1:], stack)
        return
    elif input[0] == "]":
        assert stack[-1] == "["
        return input[0] + state_post_value(input[1:], stack[:-1])
    elif input[0] == "}":
        assert stack[-1] == "{"
        return input[0] + state_post_object(input[1:], stack[:-1])

@state
def state_value_string(input: str, stack: list[str]) -> str:
    if input[0] == '"':
        # TODO: 
        return input[0] + state_post_value(input[1:], stack)

    if input[0] == "\\":
        return input[0] + state_value_string(input[1:], stack + ["v"])
    
    return input[0] + state_value_string(input[1:], stack)
    
def state_property_string(input: str, stack: list[str]) -> str:
    if input[0] == '"':
        return input[0] + state_post_property(input[1:], stack)

    if input[0] == "\\":
        return input[0] + state_property_string(input[1:], stack + ["p"])
    
    return input[0] + state_property_string(input[1:], stack)

@state
def state_escape_char(input: str, stack: list[str]) -> str:
    if input[0] == "u":
        int(input[1:5], 16)
    
        if stack[-1] == "v":
            return input[0:5] + state_value_string(input[5:], stack[:-1]) 
        if stack[-1] == "p":
            return input[0] + state_property_string(input[5:], stack[:-1])

        return        
    if input[0] in {"\\", "/", '"', "b", "f", "n", "r", "t"}:
        if stack[-1] == "v":
            return input[0] + state_value_string(input[1:], stack[:-1])
        if stack[-1] == "p":
            return input[0] + state_property_string(input[1:], stack[:-1])
        return

@state
def state_int(input: str, stack: list[str]) -> str:
    if input[0].isdigit():
        return input[0] + state_int(input[1:], stack)
    if input[0] == ".":
        return input[0] + state_double(input[1:], stack)
    if input[0] == ",":
        if stack[-1] == "[":
            return input[0] + state_value(input[1:], stack)
        if stack[-1] == "{":
            return input[0] + state_object(input[1:], stack)
        return
    if input[0] == "}":
        assert stack[-1] == "{"
        return input[0] + state_post_int_parent(input[1:], stack)
    if input[0] == "]":
        assert stack[-1] == "["
        return input[0] + state_post_int_parent(input[1:], stack)
    if input[0].isspace():
        return input[0] + state_post_value(input[1:], stack)
    
@state
def state_post_int_parent(input: str, stack: list[str]) -> str:
    if stack[-1] == "[":
        return state_post_value(input, stack[:-1])
    if stack[-1] == "{":
        return state_post_object(input, stack[:-1])
    
@state
def state_double(input: str, stack: list[str]) -> str:
    if input[0].isdigit():
        return input[0] + state_double(input[1:], stack)
    if input[0] == ",":
        if stack[-1] == "[":
            return input[0] + state_value(input[1:], stack)
        if stack[-1] == "{":
            return input[0] + state_object(input[1:], stack)
        return
    if input[0] == "}":
        assert stack[-1] == "{"
        return input[0] + state_post_int_parent(input[1:], stack)
    if input[0] == "]":
        assert stack[-1] == "["
        return input[0] + state_post_int_parent(input[1:], stack)
    if input[0].isspace():
        return input[0] + state_post_value(input[1:], stack)
    if input[0] == {"e", "E"}:
        return input[0] + state_exponent_sign(input[1:], stack)
    
@state
def state_exponent_sign(input: str, stack: list[str]) -> str:
    if input[0] in {"+", "-"}:
        return input[0] + state_exponent_digits(input[1:], stack)

@state
def state_exponent_digits(input: str, stack: list[str]) -> str:
    if input[0].isdigit():
        return input[0] + state_exponent_digits(input[1:], stack)
    if input[0] == ",":
        if stack[-1] == "[":
            return input[0] + state_value(input[1:], stack)
        if stack[-1] == "{":
            return input[0] + state_object(input[1:], stack)
        return
    if input[0] == "}":
        assert stack[-1] == "{"
        return input[0] + state_post_int_parent(input[1:], stack)
    if input[0] == "]":
        assert stack[-1] == "["
        return input[0] + state_post_int_parent(input[1:], stack)
    if input[0].isspace():
        return input[0] + state_post_value(input[1:], stack)
    

def correct_json(json_str: str) -> str:
    return state_start(json_str)