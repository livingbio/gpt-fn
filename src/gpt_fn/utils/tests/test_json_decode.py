import json
from pathlib import Path
from typing import Any

import pytest
from syrupy.assertion import SnapshotAssertion

from ..json_decode import correct_json, loads


def load_fixed_json(json_str: str) -> dict[str, Any]:
    return json.loads(correct_json(json_str))


def test_fixed_json_simple_case(snapshot: SnapshotAssertion) -> None:
    assert snapshot == load_fixed_json("{}")
    assert snapshot == load_fixed_json('{"foo": "bar"}')
    assert snapshot == load_fixed_json('{"foo": "bar", "baz": "qux"}')
    assert snapshot == load_fixed_json('{"foo": "bar", "baz": "qux", "quux": "corge"}')
    # add more nested objects
    assert snapshot == load_fixed_json('{"foo": {"bar": "baz"}}')
    assert snapshot == load_fixed_json('{"foo": {"bar": {"baz": "qux"}}}')
    assert snapshot == load_fixed_json('{"foo": {"bar": {"baz": {"qux": "quux"}}}}')


@pytest.mark.parametrize("test_filename", (Path(__file__).parent / "test_json_decode").glob("safe/*.json"), ids=lambda x: x.name)
def test_fixed_json_safe_case(snapshot: SnapshotAssertion, test_filename: Path) -> None:
    content = test_filename.read_text()
    result = load_fixed_json(content)
    assert snapshot == result
    assert result == json.loads(content)


@pytest.mark.parametrize("test_filename", (Path(__file__).parent / "test_json_decode").glob("unsafe/*.jsonx"), ids=lambda x: x.name)
def test_fixed_json_unsafe_case(snapshot: SnapshotAssertion, test_filename: Path) -> None:
    content = test_filename.read_text()
    result = load_fixed_json(content)
    assert snapshot == result


def test_fixed_json_fail() -> None:
    with pytest.raises(json.decoder.JSONDecodeError) as e:
        # test that it will raise JSONDecodeError if it can't fix the JSON
        loads("{", autofix=True)
