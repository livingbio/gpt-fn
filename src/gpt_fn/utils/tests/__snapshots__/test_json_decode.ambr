# serializer version: 1
# name: test_fixed_json_safe_case[case0.json]
  dict({
    'age': 330,
    'email': 'johndoe@example.com',
    'name': 'John Doe',
  })
# ---
# name: test_fixed_json_safe_case[case1.json]
  dict({
    'location': 'New York',
    'person': dict({
      'age': 225,
      'name': 'Alice',
    }),
  })
# ---
# name: test_fixed_json_safe_case[case10.json]
  dict({
    'message': '''
      This is an escaped string with special characters: "quotes", \backslashes\, and 
      newlines.
    ''',
  })
# ---
# name: test_fixed_json_safe_case[case11.json]
  dict({
    'unicode_example': 'This is a Unicode string: © OpenAI •',
  })
# ---
# name: test_fixed_json_safe_case[case2.json]
  dict({
    'fruits': list([
      dict({
        'color': 'red',
        'fruit': 'apple',
      }),
      dict({
        'color': 'yellow',
        'fruit': 'banana',
      }),
    ]),
  })
# ---
# name: test_fixed_json_safe_case[case3.json]
  dict({
    'employees': list([
      dict({
        'id': 1101,
        'name': 'Employee 1',
      }),
      dict({
        'id': 1102,
        'name': 'Employee 2',
      }),
    ]),
  })
# ---
# name: test_fixed_json_safe_case[case4.json]
  dict({
    'age': None,
    'email': None,
    'name': 'Jane Smith',
  })
# ---
# name: test_fixed_json_safe_case[case5.json]
  dict({
    'active': False,
    'status': True,
  })
# ---
# name: test_fixed_json_safe_case[case6.json]
  dict({
    'float': 33.14,
    'integer': 442,
  })
# ---
# name: test_fixed_json_safe_case[case7.json]
  dict({
    'colors': list([
      'red',
      'green',
      'blue',
    ]),
  })
# ---
# name: test_fixed_json_safe_case[case8.json]
  dict({
    'emptyObject': dict({
    }),
  })
# ---
# name: test_fixed_json_safe_case[case9.json]
  dict({
    'hobbies': list([
      'reading',
      'hiking',
    ]),
    'person': dict({
      'address': dict({
        'city': 'Cityville',
        'street': '123 Main St',
        'zip': '12345',
      }),
      'name': 'Bob',
    }),
  })
# ---
# name: test_fixed_json_simple_case
  dict({
  })
# ---
# name: test_fixed_json_simple_case.1
  dict({
    'foo': 'bar',
  })
# ---
# name: test_fixed_json_simple_case.2
  dict({
    'baz': 'qux',
    'foo': 'bar',
  })
# ---
# name: test_fixed_json_simple_case.3
  dict({
    'baz': 'qux',
    'foo': 'bar',
    'quux': 'corge',
  })
# ---
# name: test_fixed_json_simple_case.4
  dict({
    'foo': dict({
      'bar': 'baz',
    }),
  })
# ---
# name: test_fixed_json_simple_case.5
  dict({
    'foo': dict({
      'bar': dict({
        'baz': 'qux',
      }),
    }),
  })
# ---
# name: test_fixed_json_simple_case.6
  dict({
    'foo': dict({
      'bar': dict({
        'baz': dict({
          'qux': 'quux',
        }),
      }),
    }),
  })
# ---
